include/dsn/utility/ringbuf.h

```diff
@@ -0,0 +1,104 @@
+// Copyright (c) 2017, Xiaomi, Inc.  All rights reserved.
+// This source code is licensed under the Apache License Version 2.0, which
+// can be found in the LICENSE file in the root directory of this source tree.
+
+#pragma once
+
+/*
+Example：
+
+int main()
+{
+    ringbuf<int, 5> q;
+    for (int i=1;i<=7;i++){
+        q.push(i);
+        std::cout<<q.front()<<" "<<q.back()<<" "<<q.size()<<std::endl;
+    }
+    std::cout<<"-------"<<std::endl;
+    while (!q.empty()){
+        std::cout<<q.front()<<" "<<q.back()<<" "<<q.size()<<std::endl;
+        q.pop();
+    }
+}
+
+Result:
+1 1 1
+1 2 2
+1 3 3
+1 4 4
+1 5 5
+2 6 5
+3 7 5
+-------
+3 7 5
+4 7 4
+5 7 3
+6 7 2
+7 7 1
+
+*/
+
+namespace dsn {
+namespace utils {
+
+template <typename Type, int MaxElements>
+class ringbuf
+{
+public:
+    ringbuf() : _size(0), _head(0), _tail(0) {}
+
+    void push(const Type &obj)
+    {
+        _size++;
+        _size = std::min(_size, MaxElements);
+        if (_tail == _head && _size == MaxElements) {
```

Comment:
When `_size == MaxElements`, `_tail == _head ` is definitely true, right?

Reply:
Yes